tests/test_aes.py

```diff
@@ -123,3 +124,127 @@ def test_wrap(self):
 
         self.assertEqual(key[pkcs11.Attribute.VALUE],
                          key2[pkcs11.Attribute.VALUE])
+
+    @parameterized.expand([
+        ("POSITIVE_128_BIT",            128, 16),
+        ("POSITIVE_128_BIT_LONG_IV",    128, 32),
+        ("NEGATIVE_128_BIT_BAD_IV",     128, 15),
+        ("POSITIVE_256_BIT_LONG_IV",    256, 32),
+        ("NEGATIVE_256_BIT_SHORT_IV",   256, 16),
+        ("NEGATIVE_256_BIT_BAD_IV",     256, 31),
+    ])
+    @requires(Mechanism.AES_ECB_ENCRYPT_DATA)
+    @FIXME.opencryptoki  # can't set key attributes
+    def test_derive_ecb_encrypt(self, test_type, test_key_length, iv_length):
+        """
+        Function to test AES Key Derivation using the ECB_ENCRYPT Mechanism.
+
+        Refer to Section 2.15 of http://docs.oasis-open.org/pkcs11/pkcs11-curr/v2.40/errata01/os/pkcs11-curr-v2.40-errata01-os-complete.html#_Toc441850521
+        """
+
+        # Create the Master Key
+        capabilities = pkcs11.defaults.DEFAULT_KEY_CAPABILITIES[pkcs11.KeyType.AES]
+        capabilities |= pkcs11.MechanismFlag.DERIVE
+        key = self.session.generate_key(pkcs11.KeyType.AES, key_length=test_key_length,
+                                        capabilities=capabilities,
+                                        template={
+                                            pkcs11.Attribute.EXTRACTABLE: True,
+                                            pkcs11.Attribute.DERIVE: True,
+                                            pkcs11.Attribute.SENSITIVE: False,
+                                        })
+
+        self.assertTrue(key is not None, f"Failed to create {test_key_length}-bit Master Key")
+
+        # Derive a Key from the Master Key
+        iv = b'0' * iv_length
+        try:
+            derived_key = key.derive_key(pkcs11.KeyType.AES, key_length=test_key_length,
+                                         capabilities=capabilities,
+                                         mechanism=Mechanism.AES_ECB_ENCRYPT_DATA,
+                                         mechanism_param=iv,
+                                         template={
+                                             pkcs11.Attribute.EXTRACTABLE: True,
+                                             pkcs11.Attribute.SENSITIVE: False,
+                                         })
+        except (pkcs11.exceptions.MechanismParamInvalid,
+                pkcs11.exceptions.FunctionFailed) as e:
+            derived_key = None
+
+        if test_type.startswith("NEGATIVE"):
```

Comment:
Split these into two separate tests.

Reply:
Although, now I think about it, do we honestly need to parameterize so many tests? It feels here like we're actually testing the PKCS#11 library, not the wrapper.

Reply:
Essentially these tests validate what "should" work with PKCS11.
I agree that they might be doing more than just verify our changes, but I felt that these tests might also serve as reference points of what works vs what shall not work. this might be useful for any other newbie developers (like me).
I would still remove them, if you like.
in the meantime, I would upload the test split across key_derivation-only and key_derivation+data_enc/dec
I await your feedback :-)

Reply:
I'm okay to keep the parameterised, I just don't like branches in tests. It means you effectively now need to be coverage testing your tests to know if you got it all.

Consider passing in `func` which can either be `assertIsNone` or `assertIsNotNone` and not having a branch at all. `func` might have to be a string that calls `getattr(self, func)(...)`, but that will still fail hard if the function doesn't exist.